docs/design/RUC-on-type.md

```diff
@@ -0,0 +1,189 @@
+static void TestRequiresOnBaseButNotOnDerived ()
+{
+    DerivedWithoutRequires.StaticMethodInInheritedClass ();
+    DerivedWithoutRequires.StaticMethod ();
+}
+```
+As the previous example ilustrates the DerivedType without RequiresUnreferencedCode will generate a warning only for deriving from ClassWithRequiresUnreferencedCode. Additionally, by calling some of the methods we can see the difference on behavior on methods declared on the derived type vs methods on the base type. Calling a method in the derived type will not generate a warning since the derived type is not annotated, while calling a method for which the base type will be called will generate IL2026.
+
+Arguably we could say that the warning in the type declaration is not necessary and that we could rely on the warning being generated in the constructor instead, but this could lead to some behaviors in which we can call a method without ever using the constructor.
+
+```C#
+var t = Type.GetType("Derived").GetMethod("SomeMethod").Invoke(null, null);
+
+[RequiresUnreferencedCode("This base type is dangerous")]
+class Base {}
+
+class Derived : Base 
+{
+    static void SomeMethod() {}
+}
+```
+Therefore, we decided that generating a warning in type declaration even if there is no apparent use it's the correct behavior.
+
+## Behavior of RequiresUnreferencedCode on type while used in generic instantiation
+The behavior of the RequiresUnreferencedCode on a type when used in generics is defined on the new() constrain. Which will call the constructor of the type passed.
+```C#
+[RequiresUnreferencedCode ("Message for --ClassWithRequiresUnreferencedCode--")]
+class ClassWithRequiresUnreferencedCode { }
+
+private class GenericClass1<T> where T : ClassWithRequiresUnreferencedCode, new() { }
+private class GenericClass2<T> where T : new() { }
+private class GenericClass3<T> { }
+
+static void TestRequiresUsedAsGeneric
+{
+    var classGeneric1 = new GenericClass1<ClassWithRequiresUnreferencedCode> ();
+    var classGeneric2 = new GenericClass2<ClassWithRequiresUnreferencedCode> ();
+    var classGeneric3 = new GenericClass3<ClassWithRequiresUnreferencedCode> ();
+}
+```
+For scenarios in GenericClass1 and GenericClass2 the new constraint is present which means that the constructor of the ClassWithRequiresUnreferencedCode will be called and a warning with IL2026 will be produced. In the GenericClass3 the new constraint is not present therefore the warning will not be produced.
+
+## Structs and interfaces
+We have decided to exclude to do any analysis if structs or interfaces are inside of a class marked with RequiresUnreferencedCode, this is to follow the targets of the RequiresUnreferencedCode attribute which at this moment they dont support the attribute on a struct or an interface.
```

Comment:
```suggestion
We have decided not to allow `RequiresUnreferencedCode` on structs and interfaces because the semantics is hard to guard - both structs and interfaces can be "instantiated" without running an explicit constructor and at that point all instance methods on them are callable, even indirectly through virtual dispatch.
```